helix-core/src/uri.rs

```diff
@@ -0,0 +1,102 @@
+use std::path::{Path, PathBuf};
+
+/// A generic pointer to a file location.
+///
+/// Currently this type only supports paths to local files.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[non_exhaustive]
+pub enum Uri {
+    Path(PathBuf),
+}
+
+impl Uri {
+    pub fn from_file_path<P: AsRef<Path>>(path: P) -> Self {
+        Self::Path(path.as_ref().to_path_buf())
+    }
+
+    // This clippy allow mirrors url::Url::from_file_path
+    #[allow(clippy::result_unit_err)]
+    pub fn to_url(&self) -> Result<url::Url, ()> {
+        match self {
+            Uri::Path(path) => url::Url::from_file_path(path),
+        }
+    }
+}
+
+impl From<Uri> for PathBuf {
+    fn from(uri: Uri) -> Self {
+        match uri {
+            Uri::Path(path) => path,
+        }
+    }
+}
+
+impl AsRef<Path> for Uri {
+    fn as_ref(&self) -> &Path {
+        match self {
+            Uri::Path(path) => path,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum UrlConversionError {
+    UnsupportedScheme(url::Url),
+    UnableToConvert(url::Url),
+}
+
+impl std::fmt::Display for UrlConversionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::UnsupportedScheme(url) => {
+                write!(f, "unsupported scheme in URL: {}", url.scheme())
+            }
+            Self::UnableToConvert(url) => write!(f, "unable to convert URL to file path: {}", url),
+        }
+    }
+}
+
+impl std::error::Error for UrlConversionError {}
+
+impl TryFrom<url::Url> for Uri {
+    type Error = UrlConversionError;
+
+    fn try_from(url: url::Url) -> Result<Self, Self::Error> {
+        if url.scheme() == "file" {
+            url.to_file_path()
+                .map(|path| Self::Path(helix_stdx::path::normalize(path)))
+                .map_err(|_| UrlConversionError::UnsupportedScheme(url))
+        } else {
+            Err(UrlConversionError::UnsupportedScheme(url))
+        }
+    }
+}
+
+impl TryFrom<&url::Url> for Uri {
+    type Error = UrlConversionError;
+
+    fn try_from(url: &url::Url) -> Result<Self, Self::Error> {
+        if url.scheme() == "file" {
+            url.to_file_path()
+                .map(|path| Self::Path(helix_stdx::path::normalize(path)))
+                .map_err(|_| UrlConversionError::UnsupportedScheme(url.clone()))
```

Comment:
shouldn't this use the other variant?
```suggestion
                .map_err(|_| UrlConversionError::UnableToConvert(url.clone()))
```